classic/selection/FeatureModel.js

```diff
@@ -0,0 +1,327 @@
+     * @cfg {Boolean}
+     */
+    mapSelection: false,
+
+    /**
+     * The default style for the selected features.
+     * @cfg {ol.style.Style}
+     */
+    selectStyle: new ol.style.Style({
+        image: new ol.style.Circle({
+            radius: 6,
+            fill: new ol.style.Fill({
+                color: 'rgba(255,255,255,0.8)'
+            }),
+            stroke: new ol.style.Stroke({
+                color: 'darkblue',
+                width: 2
+            })
+        }),
+        fill: new ol.style.Fill({
+            color: 'rgba(255,255,255,0.8)'
+        }),
+        stroke: new ol.style.Stroke({
+            color: 'darkblue',
+            width: 2
+        })
+    }),
+
+    /**
+     * Lookup to preserve existing feature styles. Used to restore feature style
+     * when select style is removed.
+     * @private
+     * @property {Object}
+     */
+    existingFeatStyles: {},
+
+    /**
+     * The attribute key to mark an OL feature as selected.
+     * @cfg {String}
+     * @property
+     * @readonly
+     */
+    selectedFeatureAttr: 'gx_selected',
+
+    /**
+     * Indicates if a map click handler has been registered on init.
+     * @private
+     * @property {Boolean}
+     */
+    mapClickRegistered: false,
+
+    /**
+     * Prepare several connected objects once the selection model is ready.
+     *
+     * @private
+     */
+    bindComponent: function() {
+        var me = this;
+
+        me.callParent(arguments);
+
+        me.selectedFeatures = new ol.Collection();
+
+        // detect a layer from the store if not passed in
+        if (!me.layer || !me.layer instanceof ol.layer.Vector) {
+            var store = me.getStore();
+            if (store && store.getLayer && store.getLayer() &&
+                store.getLayer() instanceof ol.layer.Vector) {
+                me.layer = store.getLayer();
+            }
+        }
+
+        if (!me._destroying) {
+            // bind several OL events since this is not called while destroying
+            me.bindOlEvents();
+        }
+    },
+
+    /**
+     * Binds several events on the OL objects used in this class.
+     *
+     * @private
+     */
+    bindOlEvents: function() {
+        var me = this;
+
+        // change style of selected feature
+        me.selectedFeatures.on('add', me.onSelectFeatAdd, me);
+
+        // reset style of no more selected feature
+        me.selectedFeatures.on('remove', me.onSelectFeatRemove, me);
+
+        // create a map click listener for connected vector layer
+        if (me.mapSelection && me.layer && me.map) {
+            me.map.on('singleclick', me.onFeatureClick, me);
+            me.mapClickRegistered = true;
+        }
+    },
+
+    /**
+     * Unbinds several events that were registered on the OL objects in this
+     * class (see #bindOlEvents).
+     *
+     * @private
+     */
+    unbindOlEvents: function() {
+        var me = this;
+
+        if (me.selectedFeatures) {
```

Comment:
~~Shouldn't we unconditionally unbind? Really unsure.~~ Stupid me, leave it as it is.